### PR TITLE
bioptim exo PR

### PR DESCRIPTION
**Pull request by AurelienRenou**

### CODE REVIEW

File: tracking/tracking_ocp.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        self.biorbd_model = biorbd.Model(self.model_path)
+
+        self.q_file = os.path.splitext(c3d_path)[0] + "_q.txt"
+        self.qdot_file = os.path.splitext(c3d_path)[0] + "_qdot.txt"
+
+        self.data = C3dData(c3d_path, self.biorbd_model)
+
+        self.final_time = self.data.get_final_time() if final_time is None else final_time
+        self.n_shooting_points = n_shooting_points
+        self.nb_iteration = nb_iteration
```
Comment (Ipuch) [inline tracking/tracking_ocp.py:94]: put this higher in the script

File: tracking/tracking_ocp.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    def prepare_ocp(self):
+        nb_q = self.biorbd_model.nbQ()
+        nb_qdot = self.biorbd_model.nbQdot()
+
+        # Add objective functions
+        objective_functions = ObjectiveList()
+        objective_functions.add(ObjectiveFcn.Lagrange.MINIMIZE_CONTROL, key="tau", weight=10)
+        objective_functions.add(ObjectiveFcn.Mayer.TRACK_MARKERS, weight=100000, target=self.markers_ref[0], node=Node.ALL)
+        if self.with_floating_base:
+            objective_functions.add(ObjectiveFcn.Lagrange.MINIMIZE_STATE, weight=10, key="qdot",
```
Comment (Ipuch) [inline tracking/tracking_ocp.py:113]: weight to lower maybe 1

File: tracking/tracking_ocp.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                                    index=[0, 1, 2, 3, 4, 5])
```
Comment (Ipuch) [inline tracking/tracking_ocp.py:114]: Remove indexes. I think we can remove them, and the if also. It doesn't matter if we are in a floating base or not.

File: tracking/tracking_ocp.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            dynamics=dynamics,
+            n_shooting=self.n_shooting_points,
+            phase_time=self.final_time,
+            x_init=x_init,
+            u_init=u_init,
+            x_bounds=x_bounds,
+            u_bounds=u_bounds,
+            objective_functions=objective_functions,
+            ode_solver=self.ode_solver,
+            n_threads=6
```
Comment (Ipuch) [inline tracking/tracking_ocp.py:156]: Please put n_thread has an argument and an attribute of the object, by default 6,

File: tracking/tracking_ocp.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        )
+
+    def plot_c3d_markers_and_model_markers(self, c3d_path):
```
Comment (Ipuch) [inline tracking/tracking_ocp.py:159]: hints and docstrings

File: tracking/viz_tracking.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+"""
+This file contains functions to plot extra information for the optimal control problem of the Miller.
+"""
+
+import numpy as np
+import biorbd_casadi as biorbd
+from bioptim import PlotType, NonLinearProgram, OptimalControlProgram
+
+
+def marker_model(x, nlp: NonLinearProgram, idx):
```
Comment (Ipuch) [inline tracking/viz_tracking.py:10]: update hints and docstrings

File: tracking/viz_tracking.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    Compute the angular momentum of the system.
+
+    Parameters
+    ----------
+    x
+        State vector
+    nlp: NonLinearProgram
+        Non linear program
+    """
```
Comment (Ipuch) [inline tracking/viz_tracking.py:43]: Same docstrings and hints

**Review by Ipuch — COMMENTED** [review]
It starts to be something, Good job :)
Let's merge this by the end of the week to make more specific PR.

File: data/inverse_kinematics.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    dof = model.nameDof()[global_dof].to_string().replace(seg_name_new, "")
+    dof = dof.replace("_", "")  # we remove the _ "_DofName"
+    count_dof = 0
+    while model.segment(seg_id).nameDof(count_dof).to_string() != dof:
+        count_dof += 1
+
+    return seg_id, count_dof
+
+
+def apply_offset(biorbd_model, dof_nb, offset):
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:129]: done

File: data/inverse_kinematics.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    -------
+    """
+    seg_id, dof_id = get_segment_and_dof_id_from_global_dof(biorbd_model, dof_nb)
+    range_min = model.segment(seg_id).QRanges()[dof_id].min()
+    range_max = model.segment(seg_id).QRanges()[dof_id].max()
+    value = q_recons[dof_nb, :].mean()
+    delta_range_max = value - range_max
+    delta_range_min = value - range_min
+
+    offset = np.pi
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:152]: done, i remove this one

File: data/inverse_kinematics.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    count_dof = 0
+    while model.segment(seg_id).nameDof(count_dof).to_string() != dof:
+        count_dof += 1
+
+    return seg_id, count_dof
+
+
+def apply_offset(biorbd_model, dof_nb, offset):
+    """
+    Allow the users to apply an offset on a dof in the model if this dof is out of is range
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:131]: done

File: data/inverse_kinematics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:129]: done

File: data/inverse_kinematics.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                plt.plot(q_old[count, :], label=f"{str(count)} {old}")
+                plt.plot(q[count, :], label=str(count))
+                plt.plot([biorbd_model.segment(i).QRanges()[j].min()] * q.shape[1])
+                plt.plot([biorbd_model.segment(i).QRanges()[j].max()] * q.shape[1])
+                count += 1
+                plt.legend()
+    plt.show()
+
+
+def get_segment_and_dof_id_from_global_dof(biorbd_model, global_dof):
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:89]: done

File: tracking/viz_tracking.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            # f"{nlp[0].model.segment(id_marker_1).name().to_string()}",
+            f"{'Marker'} {id_marker}",
+            lambda t, x, u, p: marker_model(x, nlp[0], id_marker),
+            legend=[f"{'Marker'} {id_marker} {'x'}",
+                    f"{'Marker'} {id_marker} {'y'}",
+                    f"{'Marker'} {id_marker} {'z'}"],
+            plot_type=PlotType.PLOT,
+            node_idx=None,
+        )
+    return ocp
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/viz_tracking.py]: done

File: data/inverse_kinematics.py [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+try:
+    import bioviz
+
+    biorbd_viz_found = True
+except ModuleNotFoundError:
+    biorbd_viz_found = False
+
+
+def get_unit_division_factor(c3d_file):
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply data/inverse_kinematics.py:30]: done

File: models/KINOVA_merge.biomod [hunk 14]
```diff
@@ -2,7 +2,7 @@ version 4
 
 // File extracted from C:/Users/remi6/Desktop/S5/Wu_shoulder_model/Wu_Shoulder_Model_kinova_definitif.osim
 
-gravity	0 -9.8000000000000007 0
+gravity	0 -9.81 0
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/KINOVA_merge.biomod:5]: done

File: models/KINOVA_merge.biomod [hunk 15]
```diff
@@ -121,8 +121,9 @@ gravity	0 -9.8000000000000007 0
             -0.6406235    -0.767855146    -6.9262e-05    0
             0.542796665    -0.452920435    0.707272832    0
             0    0    0    1
-        rotations z
-        ranges		-1.570796326795	1.570796326795
+        // rotations z
+        // ranges		-1.570796326795	1.570796326795
+        // 13 april 2022 : We removed this rotation which is unpratictal for inverse kinematics using kalman
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/KINOVA_merge.biomod:126]: done

File: models/KINOVA_merge.biomod [hunk 16]
```diff
@@ -242,7 +243,7 @@ gravity	0 -9.8000000000000007 0
             0.0001    0.97951422    0.20137498    0
             0    0    0    1
         rotations x
-        ranges		-1.9	0.35
+        ranges		-1.92	1.5
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/KINOVA_merge.biomod:246]: done

File: models/KINOVA_merge.biomod [hunk 17]
```diff
@@ -266,7 +267,8 @@ gravity	0 -9.8000000000000007 0
             0.060042519    0.959314972    -0.275879829    0
             0    0    0    1
         rotations z
-        ranges		0.0	3.0
+        // ranges		0.0	3.0
+        ranges		-1.9	0.35
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/KINOVA_merge.biomod:271]: done

File: models/KINOVA_merge.biomod [hunk 18]
```diff
@@ -347,7 +349,8 @@ gravity	0 -9.8000000000000007 0
             -0.006469738    -0.114683666    0.993380994    0
             0    0    0    1
         rotations z
-        ranges		0.0	2.268928027593
+        //ranges		0.0	2.268928027593
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/KINOVA_merge.biomod:352]: done

File: models/wu_converted_definitif.bioMod [hunk 19]
```diff
@@ -327,7 +312,8 @@ gravity	0 -9.81 0
             0.060042519    0.959314972    -0.275879829    0
             0    0    0    1
         rotations z
-        ranges		0.0	3.0
+        //ranges		0.0	3.0
+        ranges		-1.9	0.35
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/wu_converted_definitif.bioMod:316]: done

File: models/wu_converted_definitif.bioMod [hunk 20]
```diff
@@ -427,7 +413,8 @@ gravity	0 -9.81 0
             -0.006469738    -0.114683666    0.993380994    0
             0    0    0    1
         rotations z
-        ranges		0.0	2.268928027593
+        //ranges		0.0	2.268928027593
+        ranges		-pi	pi
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply models/wu_converted_definitif.bioMod:417]: done

File: tracking/Animate_my_solution.py [hunk 21]
```diff
@@ -0,0 +1,17 @@
+from bioptim import OptimalControlProgram
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/Animate_my_solution.py:1]: done

File: tracking/load_experimental_data.py [hunk 22]
```diff
@@ -5,33 +5,31 @@
...
+       is_direct_shooting: bool
+           indicating if the ode solver is direct shooting method
+       Methods
+       -------
+       integrator(self, ocp, nlp) -> list
+           The interface of the OdeSolver to the corresponding integrator
+       prepare_dynamic_integrator(ocp, nlp)
+           Properly set the integration in an nlp
+       """
+    def __init__(self, file_path, biorbd_model):
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py:31]: done

File: tracking/load_experimental_data.py [hunk 23]
```diff
@@ -5,33 +5,31 @@
 
 
 class C3dData:
-    def __init__(self, file_path):
+    """
+       The base class for the ODE solvers
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py]: done

File: tracking/load_experimental_data.py [hunk 24]
```diff
@@ -5,33 +5,31 @@
...
+           indicating if the ode solver is direct collocation method
+       is_direct_shooting: bool
+           indicating if the ode solver is direct shooting method
+       Methods
+       -------
+       integrator(self, ocp, nlp) -> list
+           The interface of the OdeSolver to the corresponding integrator
+       prepare_dynamic_integrator(ocp, nlp)
+           Properly set the integration in an nlp
+       """
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py]: done

File: tracking/load_experimental_data.py [hunk 25]
```diff
@@ -72,20 +69,20 @@ def get_final_time(self):
 
 
 class LoadData:
-    def __init__(self, model, c3d_file):
+    def __init__(self, model, c3d_file, q_file, qdot_file):
         def load_txt_file(file_path, size):
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py:74]: done

File: tracking/load_experimental_data.py [hunk 26]
```diff
@@ -72,20 +69,20 @@ def get_final_time(self):
...
         self.nb_markers = model.nbMarkers()
 
         # files path
-        self.c3d_data = C3dData(c3d_file)
+        self.c3d_data = C3dData(c3d_file, model)
+        self.q = load_txt_file(q_file, self.nb_q)
+        self.qdot = load_txt_file(qdot_file, self.nb_qdot)
 
     def dispatch_data(self, data, nb_shooting: list, phase_time: list):
         """
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py:89]: done

File: tracking/load_experimental_data.py [hunk 27]
```diff
@@ -96,19 +93,27 @@ def dispatch_data(self, data, nb_shooting: list, phase_time: list):
...
-            f = interp1d(t_init, x, kind="cubic")
+            f = interp1d(t_init, x, kind="linear")
             out.append(f(t_node))
         return out
 
-    def get_marker_ref(self,
-            nb_shooting: list,
-            phase_time: list,
-            type: str) -> list:
-        #todo: add an argument if "all" all markers and if "hand" only markers of hand if "MET5" only MET5
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py:112]: done

File: tracking/load_experimental_data.py [hunk 28]
```diff
@@ -96,19 +93,27 @@ def dispatch_data(self, data, nb_shooting: list, phase_time: list):
...
+    def get_marker_ref(self, nb_shooting: list, phase_time: list, type: str) -> list:
+        # todo: add an argument if "all" all markers and if "hand" only markers of hand if "MET5" only MET5
 
         return self.dispatch_data(self.c3d_data.trajectories, nb_shooting=nb_shooting, phase_time=phase_time)
+
+    def get_experimental_data(self, number_shooting_points, phase_time, with_floating_base: bool):
+        q_ref = self.dispatch_data(data=self.q, nb_shooting=number_shooting_points, phase_time=phase_time)
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/load_experimental_data.py:112]: done

File: tracking/tracking_ocp.py [hunk 29]
```diff
@@ -0,0 +1,178 @@
...
+    QAndQDotBounds,
+    InitialGuessList,
+    OdeSolver,
+    Node,
+)
+
+
+class TrackingOcp:
+    """
+        The base class for the ODE solvers
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/tracking_ocp.py]: done

File: tracking/tracking_ocp.py [hunk 30]
```diff
@@ -0,0 +1,178 @@
...
+        model_path_with_floating_base = "../models/wu_converted_definitif.bioMod"
+        self.ode_solver = ode_solver
+        self.model_path = model_path_with_floating_base if self.with_floating_base \
+            else model_path_without_floating_base
+        self.biorbd_model = biorbd.Model(self.model_path)
+
+        self.q_file = os.path.splitext(c3d_path)[0] + "_q.txt"
+        self.qdot_file = os.path.splitext(c3d_path)[0] + "_qdot.txt"
+
+        self.data = C3dData(c3d_path, self.biorbd_model)
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/tracking_ocp.py:91]: done

File: tracking/tracking_ocp.py [hunk 31]
```diff
@@ -0,0 +1,178 @@
...
+                 c3d_path: str,
+                 n_shooting_points: int,
+                 nb_iteration: int,
+                 ode_solver: OdeSolver = OdeSolver.RK4(),
+                 final_time: float = None,
+                 ):
+        self.with_floating_base = with_floating_base
+        model_path_without_floating_base = "../models/" \
+                                           "wu_converted_definitif_without_floating_base_and_thorax_markers.bioMod"
+        model_path_with_floating_base = "../models/wu_converted_definitif.bioMod"
```
[the remark this answers is not in the document]
Comment (AurelienRenou) [reply tracking/tracking_ocp.py]: done

File: tracking/tracking_ocp.py (hunk 1 above)
Comment (AurelienRenou) [reply tracking/tracking_ocp.py:95]: done

File: tracking/tracking_ocp.py (hunk 3 above)
Comment (AurelienRenou) [reply tracking/tracking_ocp.py]: done

File: tracking/tracking_ocp.py (hunk 4 above)
Comment (AurelienRenou) [reply tracking/tracking_ocp.py]: done

File: tracking/tracking_ocp.py (hunk 5 above)
Comment (AurelienRenou) [reply tracking/tracking_ocp.py:162]: done

File: tracking/viz_tracking.py (hunk 7 above)
Comment (AurelienRenou) [reply tracking/viz_tracking.py:45]: done

File: tracking/viz_tracking.py (hunk 6 above)
Comment (AurelienRenou) [reply tracking/viz_tracking.py:10]: done

File: tracking/load_experimental_data.py [hunk 32]
```diff
@@ -5,39 +5,37 @@
...
-            "MET2",  # check 21
-            "MET5",  # check 22
-        ]
-
+        self.marker_names = [biorbd_model.markerNames()[i].to_string() for i in range(len(biorbd_model.markerNames()))]
         self.trajectories = self.get_marker_trajectories(self.c3d, self.marker_names)
 
     @staticmethod
-    def get_marker_trajectories(loaded_c3d, marker_names):
+    def get_marker_trajectories(loaded_c3d: c3d, marker_names: list) -> np.ndarray:
```
Comment (Ipuch) [inline tracking/load_experimental_data.py:35]: add self as an argument,
remove c3d and use self.c3d in the function
marker_names is None by default.
this is not a staticmethod

**Pull request merged**